backend/OKPluginVariableRendering/VariableRendering.cs

```diff
@@ -0,0 +1,379 @@
+                                if (gatheredAttributes[a.Name].Priority < a.Priority)
+                                {
+                                    gatheredAttributes[a.Name] = a;
+                                }
+                            }
+                            else
+                            {
+                                gatheredAttributes.Add(a.Name, a);
+                            }
+                        }
+                    }
+                }
+
+                foreach (var followRelation in cfg.BaseCI.FollowRelations)
+                {
+                    MergedCI prevCI = mainCI;
+
+                    // save a list with ids for first level of realtions we will need this to check to the second
+                    var relationsCIs = new Queue<MergedCI>();
+                    relationsCIs.Enqueue(mainCI);
+
+                    foreach (var follow in followRelation.Follow)
+                    {
+
+                        var tmpRelationsCIs = new List<MergedCI>();
+
+                        while (relationsCIs.Any())
+                        {
+                            prevCI = relationsCIs.Dequeue();
+
+                            // get the predicate remove the first char, first char defines the direction of relation
+                            var predicate = follow.Predicate[1..];
+
+                            // find CI relations with this predicate
+                            List<MergedRelation> mergedRelations = new();
+
+                            if (follow.Predicate[0] == '>')
+                            {
+                                if (allFromRelations.ContainsKey($"{predicate}-{prevCI.ID}"))
+                                {
+                                    mergedRelations = allFromRelations[$"{predicate}-{prevCI.ID}"];
+                                }
+                            }
+                            else
+                            {
+                                if (allFromRelations.ContainsKey($"{predicate}-{prevCI.ID}"))
+                                {
+                                    mergedRelations = allFromRelations[$"{predicate}-{prevCI.ID}"];
+                                }
+                            }
+
+                            if (!mergedRelations.Any())
+                            {
+                                continue;
+                            }
+
+                            var targetCIs = new List<MergedCI>();
+
+                            foreach (var r in mergedRelations)
+                            {
+                                var targetCI = follow.Predicate[0] == '>' ? allCIs[r.Relation.ToCIID] : allCIs[r.Relation.FromCIID];
+                                targetCIs.Add(targetCI);
+                            }
+
+                            var targetCIRequiredTrait = await traitsProvider.GetActiveTrait(follow.RequiredTrait, trans, changesetProxy.TimeThreshold);
```

Comment:
That could be really slow because its running in a hot loop. Better to fetch all active traits once at the start and use that instead of calling GetActiveTrait() that often